readme.md
# PDF Generator

Titanium module to handle PDF generation/edition on iOS using Quartz 2D for fast rendering and good quality.

# Installing

1. Download this project and open the .zip file in /build folder, then copy it's content to your project in the correct folder structure. The path will look like the following: `modules/iphone/br.com.grupow2abrasil.pdfgenerator/2.0.0`
2. Open Studio, and the `tiapp.xml` file for the project in question.
3. Switch to the `tiapp.xml` tab.
4. In the application's `tiapp.xml`, find the `<modules/>` node, and replace it with the new `<modules>` content. If you already have modules, just add a new node for the PDF Generator module. Note that the "version" and "platform" attributes are optional. When "version" is not specified, the latest version of the module will be used (as of Titanium SDK 2.0.0).
```xml
<modules>
	<module version="2.0.0" platform="iphone">br.com.grupow2abrasil.pdfgenerator</module>
</modules>
```
5. Use the require function to load the module in the app's code, for example:
```javascript
var pdfGenerator = require('br.com.grupow2abrasil.pdfgenerator');
```
6. The next time the app is launched or built, PDF Generator should be included with the application.

## Prerequisites

* Mac OS X
* Axway Appcelerator Titanium
* XCode

# API

First of all, load the module.

```javascript
var pdfGenerator = require('br.com.grupow2abrasil.pdfgenerator');
```

Then set the properties of your PDF and start drawing. To save call the savePDF function.

## Methods

All parameters with `*` are required.

### setProperties(obj`*`)
```javascript
pdfGenerator.setProperties({
    pdfName: "output",
    title: "",
    author: "",
    subject: "",
    keywords: "",
    creator: ""
});
```
Sets the properties of the PDF, *must* be called before everything if you want the properties to be setted properly.

| Name | Type | Description |
| --- |:---:| --- |
| obj | `Object` | property_name-to-property_value object structure |

### getProperties
```javascript
var props = pdfGenerator.getProperties();
```
Returns `Object` - PDF properties.

### openPDF(path`*`)
```javascript
var file = Ti.Filesystem.getFile(Ti.Filesystem.getApplicationDataDirectory(), "my_awesome_pdf.jpg");

pdfGenerator.openPDF(file.nativePath);
```
Opens a PDF where you can add new pages (`addNewPage`), delete pages (` deletePage`) and draw/write something `forEachPage`. Keep in mind that PDF is read-only, so you can not, for example, edit shapes, change texts or such things on pages that have already been finished.

| Name | Type | Description |
| --- |:---:| --- |
| path | `String` | The PDF file URL, it *must* be in the application folders. |

### addNewPage()
```javascript
pdfGenerator.addNewPage();
```
Adds a new page to the PDF.

### deletePage(page`*`)
```javascript
pdfGenerator.deletePage(3);
```
Deletes a page from the PDF.

| Name | Type | Description |
| --- |:---:| --- |
| page | `Number` | Index of the page to be deleted, starts with 1 and must be an Integer |

### forEachPage(fnc`*`)
```javascript
pdfGenerator.forEachPage(function(pageIndex, total){
	// Drawing functions here, you can draw the header and footer, example:
	pdfGenerator.drawText("Page " + pageIndex + " of " + total, 20, 700, 572, 30);

	// More drawing function as you wish
});
```
Executes a function for each page of the PDF where you can draw into.

| Name | Type | Description |
| --- |:---:| --- |
| fnc | `Function` | Function which will be executed for each page receiving the parameters index and total, both `Number`s |

### setTextColor(R`*`, G`*`, B`*`, A)
```javascript
pdfGenerator.setTextColor(27, 27, 27, 1.0);
```
Sets the text color in RGBA notation or RGB if A is omitted.

| Name | Type | Description |
| --- |:---:| --- |
| R | `Number` | Color channel value, must be a value in the range from 0 (solid black) to 255 (white) |
| G | `Number` | Color channel value, must be a value in the range from 0 (solid black) to 255 (white) |
| B | `Number`| Color channel value, must be a value in the range from 0 (solid black) to 255 (white) |
| A | `Number` | Alpha channel value, must be a value in the range from 0.0 (transparent) to 1.0 (solid color) |

### setDrawColor(R`*`, G`*`, B`*`, A)
```javascript
pdfGenerator.setDrawColor(27, 27, 27, 1.0);
```
Sets the draw color in RGBA notation or RGB if A is omitted.

| Name | Type | Description |
| --- |:---:| --- |
| R | `Number` | Color channel value, must be a value in the range from 0 (solid black) to 255 (white) |
| G | `Number` | Color channel value, must be a value in the range from 0 (solid black) to 255 (white) |
| B | `Number`| Color channel value, must be a value in the range from 0 (solid black) to 255 (white) |
| A | `Number` | Alpha channel value, must be a value in the range from 0.0 (transparent) to 1.0 (solid color) |

### setFillColor(R`*`, G`*`, B`*`, A)
```javascript
pdfGenerator.setFillColor(255, 255, 255, 0.0);
```
Sets the fill color in RGBA notation or RGB if A is omitted. This property applies only for shapes.

| Name | Type | Description |
| --- |:---:| --- |
| R | `Number` | Color channel value, must be a value in the range from 0 (solid black) to 255 (white) |
| G | `Number` | Color channel value, must be a value in the range from 0 (solid black) to 255 (white) |
| B | `Number`| Color channel value, must be a value in the range from 0 (solid black) to 255 (white) |
| A | `Number` | Alpha channel value, must be a value in the range from 0.0 (transparent) to 1.0 (solid color) |

### setLineWidth(width`*`)
```javascript
pdfGenerator.setLineWidth(0.1);
```
Sets line width for upcoming lines and shapes.

| Name | Type | Description |
| --- |:---:| --- |
| width | `Number` | Line width, must be a float/double |

### getLineWidth()
```javascript
pdfGenerator.getLineWidth();
```
Returns `Number` - Line width.

### setFontName(fontName`*`)
```javascript
pdfGenerator.setFontName("Helvetica");
```
Sets text font for upcoming text elements.

| Name | Type | Description |
| --- |:---:| --- |
| fontName | `String` | Font name or family. Style or variant must be included in the fontName, example: `Helvetica-Bold` |

### getFontName()
```javascript
pdfGenerator.getFontName();
```
Returns `String` - Current font name.

### setTextSize(size`*`)
```javascript
pdfGenerator.setTextSize(14);
```
Sets text size for upcoming text elements.

| Name | Type | Description |
| --- |:---:| --- |
| size | `Number` | Font name or family. Style or variant must be included in the fontName, example: `Helvetica-Bold` |

### getTextSize()
```javascript
pdfGenerator.getTextSize();
```
Returns `Number` - Current size of text.

### setTextAlign(alignment`*`)
```javascript
pdfGenerator.setTextAlign("center");
```
Sets text horizontal alignment for upcoming text elements.

| Name | Type | Description |
| --- |:---:| --- |
| alignment | `String` | Horizontal text alignment, available values are `left`, `center`, `right`, `justified` |

### getTextAlign()
```javascript
pdfGenerator.getTextAlign();
```
Returns `String` - Current text horizontal alignment.

### setTextVerticalAlign(alignment`*`)
```javascript
pdfGenerator.setTextVerticalAlign("middle");
```
Sets text vertical alignment for upcoming text elements.

| Name | Type | Description |
| --- |:---:| --- |
| alignment | `String` | Vertical text alignment, available values are `top`, `middle`, `bottom` |

### getTextVerticalAlign()
```javascript
pdfGenerator.getTextVerticalAlign();
```
Returns `String` - Current text vertical alignment.

### setPdfName(pdfName`*`)
```javascript
pdfGenerator.setPdfName("my_awesome_pdf");
```
Sets the file name of the generated PDF.

| Name | Type | Description |
| --- |:---:| --- |
| pdfName | `String` | The name of the PDF without file extension |

### getPdfName()
```javascript
pdfGenerator.getPdfName();
```
Returns `String` - PDF filename.

### drawText(text`*`, x`*`, y`*`, width`*`, height`*`)
```javascript
pdfGenerator.drawText("The fox jumps over the lazy dog", 20, 20, 572, 30);
```
Writes text to the PDF.

| Name | Type | Description |
| --- |:---:| --- |
| text | `String` | The text to be drawn |
| x | `Number` | Top-down horizontal coordinate where 0 is the start point at top, must be an Integer |
| y | `Number` | Left-right vertical coordinate where 0 is the start point at left, must be an Integer |
| width | `Number` | Width of the rect where the text will be drawn, must be an Integer |
| height | `Number` | Height of the rect where the text will be drawn, must be an Integer |

### drawLine(startX`*`, startY`*`, endX`*`, endY`*`)
```javascript
pdfGenerator.drawLine(20, 20, 500, 500);
```
Draws a line from one point to other.

| Name | Type | Description |
| --- |:---:| --- |
| startX | `Number` | Top-down vertical coordinate for the start point, must be an Integer |
| startY | `Number` | Left-right horizontal coordinate for the start point, must be an Integer |
| endX | `Number` | Top-down vertical coordinate for the end point, must be an Integer |
| endY | `Number` | Left-right horizontal coordinate for the end point, must be an Integer |

### drawRect(x`*`, y`*`, width`*`, height`*`)
```javascript
pdfGenerator.drawRect(10, 10, 582, 35);
```
Draws a rectangle.

| Name | Type | Description |
| --- |:---:| --- |
| x | `Number` | Top-down horizontal coordinate where 0 is the start point at top, must be an Integer |
| y | `Number` | Left-right vertical coordinate where 0 is the start point at left, must be an Integer |
| width | `Number` | Width of the rectangle, must be an Integer |
| height | `Number` | Height of the rectangle, must be an Integer |

### drawImage(path`*`, x`*`, y`*`, width`*`, height`*`)
```javascript
pdfGenerator.drawImage("\mypath\to\my\awesome\image.jpg", 20, 80, 100, 100);
```
Draws an image.

| Name | Type | Description |
| --- |:---:| --- |
| path | `String` | Image path location |
| x | `Number` | Top-down horizontal coordinate where 0 is the start point at top, must be an Integer |
| y | `Number` | Left-right vertical coordinate where 0 is the start point at left, must be an Integer |
| width | `Number` | Width of the rectangle, must be an Integer |
| height | `Number` | Height of the rectangle, must be an Integer |

### drawEllipse(x`*`, y`*`, width`*`, height`*`)
```javascript
pdfGenerator.drawEllipse(10, 10, 90, 90);
```
Draws an ellipse.

| Name | Type | Description |
| --- |:---:| --- |
| x | `Number` | Top-down horizontal coordinate where 0 is the start point at top, must be an Integer |
| y | `Number` | Left-right vertical coordinate where 0 is the start point at left, must be an Integer |
| width | `Number` | Width of the ellipse, must be an Integer |
| height | `Number` | Height of the ellipse, must be an Integer |

### addURL(url`*`, x`*`, y`*`, width`*`, height`*`)
```javascript
pdfGenerator.addURL("http://grupow2abrasil.com.br/", 20, 620, 572, 30);
```
Adds a clickable area into a rectangle which opens a link on click.

| Name | Type | Description |
| --- |:---:| --- |
| url | `String` | The link |
| x | `Number` | Top-down horizontal coordinate where 0 is the start point at top, must be an Integer |
| y | `Number` | Left-right vertical coordinate where 0 is the start point at left, must be an Integer |
| width | `Number` | Width of the rectangle for the clickable area, must be an Integer |
| height | `Number` | Height of the rectangle for the clickable area, must be an Integer |

### cancelPDF()
```javascript
pdfGenerator.cancelPDF();
```
Cancels the PDF drawing.

### savePDF(callback`*`)
```javascript
pdfGenerator.savePDF(function(e){
	Ti.API.info("My Awesome PDF is located at: " + e.url);
});
```
Saves the PDF file, calling callback when the file is ready.

| Name | Type | Description |
| --- |:---:| --- |
| pdfName | `Function` | A callback function which will be called when PDF is ready returning a dictionary with the file url |

### setPageWidth(width`*`)
```javascript
pdfGenerator.setPageWidth(620);
```
Sets PDF page width.

| Name | Type | Description |
| --- |:---:| --- |
| width | `Number` | Width of the PDF page, must be an Integer |

### getPageWidth()
```javascript
pdfGenerator.getPageWidth();
```
Returns `Number` - Page width (default is 612).

### setPageHeight(height`*`)
```javascript
pdfGenerator.setPageHeight(800);
```
Sets PDF page height.

| Name | Type | Description |
| --- |:---:| --- |
| height | `Number` | Height of the PDF page, must be an Integer |

### getPageHeight()
```javascript
pdfGenerator.getPageHeight();
```
Returns `Number` - Page height (default is 792).

### getPageCount()
```javascript
pdfGenerator.getPageCount();
```
Returns `Number` - Number of pages.

## Example

The follwing example illustrates the usage of PDF Generator.

```javascript
var docViewer = Ti.UI.iOS.createDocumentViewer();
var pdfGenerator = require("br.com.grupow2abrasil.pdfgenerator");

// The PDF name doesn't need extension .pdf, only the name
pdfGenerator.setProperties({
	pdfName: "pdfOutputName",
    title: "The title of the PDF",
    author: "The author of the PDF",
    subject: "A subject if you want",
    keywords: "Keyworkds of the PDF",
    creator: "PDF Creator"
});

pdfGenerator.addNewPage();

pdfGenerator.drawText("The fox jumps over the lazy fox.", 100, 40, 200, 40);
pdfGenerator.drawLine(10, 10, 60, 60);

pdfGenerator.drawRect(5, 520, 100, 20);

// the image file must be in documents directory
var file = Ti.Filesystem.getFile(Titanium.Filesystem.getApplicationDataDirectory(), "yourFileName.jpg");

pdfGenerator.drawImage(file.nativePath, 500, 80, 16, 16);

pdfGenerator.setTextSize(12);
pdfGenerator.drawText("I'm a text with different font size", 10, 220, 300, 40);
pdfGenerator.setTextColor(158, 67, 89);
pdfGenerator.drawText("I'm a text with different color", 10, 120, 300, 40);

pdfGenerator.setLineWidth(5.0);
pdfGenerator.drawRect(50, 80, 25, 60);

pdfGenerator.drawEllipse(200, 300, 25, 60);

pdfGenerator.savePDF(function(e){
    Ti.API.info("url: " + e.url);
    docViewer.setUrl(e.url);
    docViewer.show();
});
```

# Built With

* [Axway Appcelerator Titanium](https://www.appcelerator.com/) - Mobile Development Platform
* [XCode](https://developer.apple.com/xcode/) - iOS IDE
* [Quartz 2D](https://developer.apple.com/library/content/documentation/GraphicsImaging/Conceptual/drawingwithquartz2d/Introduction/Introduction.html) - iOS 2D Native Rendering Framework

Happy Coding!

# License

PDF Generator is licensed under MIT.

```
 Copyright (c) 2018 Grupo W2ABrasil

 Permission is hereby granted, free of charge, to any person obtaining a copy
 of this software and associated documentation files (the "Software"), to deal
 in the Software without restriction, including without limitation the rights
 to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 copies of the Software, and to permit persons to whom the Software is
 furnished to do so, subject to the following conditions:

 The above copyright notice and this permission notice shall be included in
 all copies or substantial portions of the Software.

 THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 THE SOFTWARE.
```

changelog.md
##Version 1.0.0

* First version, needs a lot of improvements and new features.

##Version 1.1.0

* Added methods `openPDF`, `cancelPDF`, `deletePage`, `forEachPage`, `getProperties`, `drawEllipse` and `addURL`
* Removed method `getPageNumbers` and improved `getPageCount` (redundant method)
* Changed `savePDF` to accept a string (the name of the pdf) as parameter instead of an object
* Added horizontal and vertical text alignments with the new properties `textAlign` and `textVerticalAlign`
* Fixed text positioning
* Revised all source code and made a lot of adjustments improving comments documentation
* Added input validation to all methods
* Fixed a lot of bugs
* Added new bugs to fix later... J/K

##Version 2.0.0

* Removed event `pdfReady` instead, now after save, a function is called
* Fixed text vertical position bug where middle was bottom instead of middle
* Few improvements
* Fixed some bugs and added new bugs to fix later

iphone/build.py
#!/usr/bin/env python
#
# Appcelerator Titanium Module Packager
#
#
import os, subprocess, sys, glob, string
import zipfile
from datetime import date

cwd = os.path.abspath(os.path.dirname(sys._getframe(0).f_code.co_filename))
os.chdir(cwd)
required_module_keys = ['architectures', 'name','version','moduleid','description','copyright','license','copyright','platform','minsdk']
module_defaults = {
	'description':'My module',
	'author': 'Your Name',
	'license' : 'Specify your license',
	'copyright' : 'Copyright (c) %s by Your Company' % str(date.today().year),
}
module_license_default = "TODO: place your license here and we'll include it in the module distribution"

def find_sdk(config):
	sdk = config['TITANIUM_SDK']
	return os.path.expandvars(os.path.expanduser(sdk))

def replace_vars(config,token):
	idx = token.find('$(')
	while idx != -1:
		idx2 = token.find(')',idx+2)
		if idx2 == -1: break
		key = token[idx+2:idx2]
		if not config.has_key(key): break
		token = token.replace('$(%s)' % key, config[key])
		idx = token.find('$(')
	return token


def read_ti_xcconfig():
	contents = open(os.path.join(cwd,'titanium.xcconfig')).read()
	config = {}
	for line in contents.splitlines(False):
		line = line.strip()
		if line[0:2]=='//': continue
		idx = line.find('=')
		if idx > 0:
			key = line[0:idx].strip()
			value = line[idx+1:].strip()
			config[key] = replace_vars(config,value)
	return config

def generate_doc(config):
	docdir = os.path.join(cwd,'documentation')
	if not os.path.exists(docdir):
		docdir = os.path.join(cwd,'..','documentation')
	if not os.path.exists(docdir):
		print "Couldn't find documentation file at: %s" % docdir
		return None

	try:
		import markdown2 as markdown
	except ImportError:
		import markdown
	documentation = []
	for file in os.listdir(docdir):
		if file in ignoreFiles or os.path.isdir(os.path.join(docdir, file)):
			continue
		md = open(os.path.join(docdir,file)).read()
		html = markdown.markdown(md)
		documentation.append({file:html});
	return documentation

def compile_js(manifest,config):
	js_file = os.path.join(cwd,'assets','br.com.grupow2abrasil.pdfgenerator.js')
	if not os.path.exists(js_file):
		js_file = os.path.join(cwd,'..','assets','br.com.grupow2abrasil.pdfgenerator.js')
	if not os.path.exists(js_file): return

	from compiler import Compiler
	try:
		import json
	except:
		import simplejson as json

	compiler = Compiler(cwd, manifest['moduleid'], manifest['name'], 'commonjs')
	root_asset, module_assets = compiler.compile_module()

	root_asset_content = """
%s

	return filterDataInRange([NSData dataWithBytesNoCopy:data length:sizeof(data) freeWhenDone:NO], ranges[0]);
""" % root_asset

	module_asset_content = """
%s

	NSNumber *index = [map objectForKey:path];
	if (index == nil) {
		return nil;
	}
	return filterDataInRange([NSData dataWithBytesNoCopy:data length:sizeof(data) freeWhenDone:NO], ranges[index.integerValue]);
""" % module_assets

	from tools import splice_code

	assets_router = os.path.join(cwd,'Classes','BrComGrupow2abrasilPdfgeneratorModuleAssets.m')
	splice_code(assets_router, 'asset', root_asset_content)
	splice_code(assets_router, 'resolve_asset', module_asset_content)

	# Generate the exports after crawling all of the available JS source
	exports = open('metadata.json','w')
	json.dump({'exports':compiler.exports }, exports)
	exports.close()

def die(msg):
	print msg
	sys.exit(1)

def warn(msg):
	print "[WARN] %s" % msg

def error(msg):
	print "[ERROR] %s" % msg

def validate_license():
	license_file = os.path.join(cwd,'LICENSE')
	if not os.path.exists(license_file):
		license_file = os.path.join(cwd,'..','LICENSE')
	if os.path.exists(license_file):
		c = open(license_file).read()
		if c.find(module_license_default)!=-1:
			warn('please update the LICENSE file with your license text before distributing')

def validate_manifest():
	path = os.path.join(cwd,'manifest')
	f = open(path)
	if not os.path.exists(path): die("missing %s" % path)
	manifest = {}
	for line in f.readlines():
		line = line.strip()
		if line[0:1]=='#': continue
		if line.find(':') < 0: continue
		key,value = line.split(':')
		manifest[key.strip()]=value.strip()
	for key in required_module_keys:
		if not manifest.has_key(key): die("missing required manifest key '%s'" % key)
		if manifest[key].strip() == '': die("manifest key '%s' missing required value" % key)
		if module_defaults.has_key(key):
			defvalue = module_defaults[key]
			curvalue = manifest[key]
			if curvalue==defvalue: warn("please update the manifest key: '%s' to a non-default value" % key)
	return manifest,path

ignoreFiles = ['.DS_Store','.gitignore','libTitanium.a','titanium.jar','README']
ignoreDirs = ['.DS_Store','.svn','.git','CVSROOT']

def zip_dir(zf,dir,basepath,ignore=[],includeJSFiles=False):
	for root, dirs, files in os.walk(dir):
		for name in ignoreDirs:
			if name in dirs:
				dirs.remove(name)	# don't visit ignored directories
		for file in files:
			if file in ignoreFiles: continue
			e = os.path.splitext(file)
			if len(e) == 2 and e[1] == '.pyc': continue
			if not includeJSFiles and len(e) == 2 and e[1] == '.js': continue
			from_ = os.path.join(root, file)
			to_ = from_.replace(dir, basepath, 1)
			zf.write(from_, to_)

def glob_libfiles():
	files = []
	for libfile in glob.glob('build/**/*.a'):
		if libfile.find('Release-')!=-1:
			files.append(libfile)
	return files

def build_module(manifest,config):
	from tools import ensure_dev_path
	ensure_dev_path()

	rc = os.system("xcodebuild -sdk iphoneos -configuration Release")
	if rc != 0:
		die("xcodebuild failed")
	rc = os.system("xcodebuild -sdk iphonesimulator -configuration Release")
	if rc != 0:
		die("xcodebuild failed")
    # build the merged library using lipo
	moduleid = manifest['moduleid']
	libpaths = ''
	for libfile in glob_libfiles():
		libpaths+='%s ' % libfile

	os.system("lipo %s -create -output build/lib%s.a" %(libpaths,moduleid))

def verify_build_arch(manifest, config):
	binaryname = 'lib%s.a' % manifest['moduleid']
	binarypath = os.path.join('build', binaryname)
	manifestarch = set(manifest['architectures'].split(' '))

	output = subprocess.check_output('xcrun lipo -info %s' % binarypath, shell=True)

	builtarch = set(output.split(':')[-1].strip().split(' '))

	print 'Check build architectures\n'

	if ('arm64' not in builtarch):
		warn('built module is missing 64-bit support.')

	if (manifestarch != builtarch):
		warn('architectures in manifest: %s' % ', '.join(manifestarch))
		warn('compiled binary architectures: %s' % ', '.join(builtarch))

		print '\nMODULE BUILD FAILED'
		error('there is discrepancy between the architectures specified in module manifest and compiled binary.')
		error('Please update manifest to match module binary architectures.')
		die('')

def package_module(manifest,mf,config):
	name = manifest['name'].lower()
	moduleid = manifest['moduleid'].lower()
	version = manifest['version']
	modulezip = '%s-iphone-%s.zip' % (moduleid,version)
	if os.path.exists(modulezip): os.remove(modulezip)
	zf = zipfile.ZipFile(modulezip, 'w', zipfile.ZIP_DEFLATED)
	modulepath = 'modules/iphone/%s/%s' % (moduleid,version)
	zf.write(mf,'%s/manifest' % modulepath)
	libname = 'lib%s.a' % moduleid
	zf.write('build/%s' % libname, '%s/%s' % (modulepath,libname))
	docs = generate_doc(config)
	if docs!=None:
		for doc in docs:
			for file, html in doc.iteritems():
				filename = string.replace(file,'.md','.html')
				zf.writestr('%s/documentation/%s'%(modulepath,filename),html)

	p = os.path.join(cwd, 'assets')
	if not os.path.exists(p):
		p = os.path.join(cwd, '..', 'assets')
	if os.path.exists(p):
		zip_dir(zf,p,'%s/%s' % (modulepath,'assets'),['README'])

	for dn in ('example','platform'):
		p = os.path.join(cwd, dn)
		if not os.path.exists(p):
			p = os.path.join(cwd, '..', dn)
		if os.path.exists(p):
			zip_dir(zf,p,'%s/%s' % (modulepath,dn),['README'],True)

	license_file = os.path.join(cwd,'LICENSE')
	if not os.path.exists(license_file):
		license_file = os.path.join(cwd,'..','LICENSE')
	if os.path.exists(license_file):
		zf.write(license_file,'%s/LICENSE' % modulepath)

	zf.write('module.xcconfig','%s/module.xcconfig' % modulepath)
	exports_file = 'metadata.json'
	if os.path.exists(exports_file):
		zf.write(exports_file, '%s/%s' % (modulepath, exports_file))
	zf.close()


if __name__ == '__main__':
	print "**************************************************************"
	print "  WARNING!"
	print "    This Python script is deprecated!"
	print "    Please use 'appc run -p ios --build-only' instead"
	print "**************************************************************"
	print ""

	manifest,mf = validate_manifest()
	validate_license()
	config = read_ti_xcconfig()

	sdk = find_sdk(config)
	sys.path.insert(0,os.path.join(sdk,'iphone'))
	sys.path.append(os.path.join(sdk, "common"))

	compile_js(manifest,config)
	build_module(manifest,config)
	verify_build_arch(manifest, config)
	package_module(manifest,mf,config)
	sys.exit(0)


iphone/Classes/BrComGrupow2abrasilPdfgeneratorModule.h
/**
 * PDF Generator
 *
 * Created by Grupo W2ABrasil
 * Copyright (c) 2018 Grupo W2ABrasil
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated
 * documentation files (the "Software"), to deal in the Software without restriction, including without limitation the
 * rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and to permit
 * persons to whom the Software is furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all copies or substantial portions of the
 * Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE
 * WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
 * COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR
 * OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

#import "TiModule.h"
#import <CoreText/CoreText.h>

@interface BrComGrupow2abrasilPdfgeneratorModule : TiModule
{
    
    CGContextRef _context;
    NSDictionary* _props;
    UIColor* _textColor;
    UIColor* _drawColor;
    UIColor* _fillColor;
}

/**
 PDF File name (without file extension), default is "output"
 */
@property(copy) NSString* pdfName;

/**
 Current font being used by drawText method, default is "Helvetica"
 */
@property(copy) NSString* fontName;

/**
 Number of pages of the PDF
 */
@property(copy) NSNumber* pageCount;

/**
 Line Width of the stroke, default is 0.1
 */
@property(copy) NSNumber* lineWidth;

/**
 Size of the text, is used by drawText method
 */
@property(copy) NSNumber* textSize;

/**
 Text alignment (left, center, right, justified), is used by drawText method
 */
@property(copy) NSString* textAlign;

/**
 Text alignment (top, middle, bottom), is used by drawText method
 */
@property(copy) NSString* textVerticalAlign;

/**
 Width of the PDF page, default is 612
 */
@property(copy) NSNumber* pageWidth;

/**
 Height of the PDF page, default is 792
 */
@property(copy) NSNumber* pageHeight;


/**
 Creates a new PDF File as copy from the specified PDF file

 @param args (path) string which is a path to the PDF file, is required
 */
-(void)openPDF:(id)args;

/**
 Sets the PDF properties

 @param args (dict) Dictionary (js object) with properties as follows:
 {
    "pdfName":"The Fox",
    "title":"John Due",
    "author":"John Due Again",
    "subject":"John is in everywhere",
    "keywords":"john, due",
    "creator":"John Due"
 }
 */
-(void)setProperties:(id)args;

/**
 Returns the PDF properties

 @param args Void
 @return return Dictionary (js object) with properties as follows
 {
     "pdfName":"The Fox",
     "title":"John Due",
     "author":"John Due Again",
     "subject":"John is in everywhere",
     "keywords":"john, due",
     "creator":"John Due"
 }
 */
-(NSDictionary *)getProperties:(id)args;


/**
 Adds a new page to the PDF

 @param args Void
 */
-(void)addNewPage:(id)args;


/**
 Deletes a page from the PDF, actually creates a new PDF without the specified page

 @param args (page) int which indicates the page to be deleted, is required
 */
-(void)deletePage:(id)args;

/**
 Creates a new PDF executing a js callback over each created page which enables redraw

 @param args (callback) KrollCallback (js callback) which will be executed for each page, is required
 */
-(void)forEachPage:(id)args;

/**
 Sets the text color for method drawText

 @param args (r, g, b, a)
 R -> Int, red color from 0 to 255, is required
 G -> Int, green color from 0 to 255, is required
 B -> Int, blue color from 0 to 255, is required
 A -> Double, alpha from 0.0 to 1.0, is optional default 1.0
 
 default color is darkGray
 */
-(void)setTextColor:(id)args;

/**
 Sets the stroke color for all drawing methods except drawText
 
 @param args (r, g, b, a)
 R -> Int, red color from 0 to 255, is required
 G -> Int, green color from 0 to 255, is required
 B -> Int, blue color from 0 to 255, is required
 A -> Double, alpha from 0.0 to 1.0, is optional default 1.0
 
 default color is darkGray
 */
-(void)setDrawColor:(id)args;

/**
 Sets the fill color for all shape drawing methods
 
 @param args (r, g, b, a)
 R -> Int, red color from 0 to 255, is required
 G -> Int, green color from 0 to 255, is required
 B -> Int, blue color from 0 to 255, is required
 A -> Double, alpha from 0.0 to 1.0, is optional default 1.0
 
 default color is darkGray
 */
-(void)setFillColor:(id)args;


/**
 Writes text into a rect

 @param args (text, x, y, width, height)
 text   -> String to be drawn, is required
 x      -> Int, the top-down horizontal coordinate where 0 is the start point at top, is required
 y      -> Int, the left-right vertical coordinate where 0 is the start point at left, is required
 width  -> Int, the width of the rect where the text will be drawn, is required
 height -> Int, the height of the rect where the text will be drawn, is required
 */
-(void)drawText:(id)args;


/**
 Draws a line between 2 points

 @param args (startX, startY, endX, endY)
 startX -> Int, the top-down vertical coordinate for the start point, is required
 startY -> Int, the left-right horizontal coordinate for the start point, is required
 endX   -> Int, the top-down vertical coordinate for the end point, is required
 endY   -> Int, the left-right horizontal coordinate for the end point, is required
 */
-(void)drawLine:(id)args;

/**
 Draws a rectangle

 @param args (x, y, width, height)
 x      -> Int, the top-down horizontal coordinate where 0 is the start point at top, is required
 y      -> Int, the left-right vertical coordinate where 0 is the start point at left, is required
 width  -> Int, the width of the rectangle, is required
 height -> Int, the height of the rectangle, is required
 */
-(void)drawRect:(id)args;

/**
 Draws an Image into a rectangle
 
 @param args (path, x, y, width, height)
 path   -> String, the path to the image, is required
 x      -> Int, the top-down horizontal coordinate where 0 is the start point at top, is required
 y      -> Int, the left-right vertical coordinate where 0 is the start point at left, is required
 width  -> Int, the width of the rectangle where the image will be drawn, is required
 height -> Int, the height of the rectangle where the image will be drawn, is required
 */
-(void)drawImage:(id)args;

/**
 Draws an Ellipse into a rectangle
 
 @param args (x, y, width, height)
 x      -> Int, the top-down horizontal coordinate where 0 is the start point at top, is required
 y      -> Int, the left-right vertical coordinate where 0 is the start point at left, is required
 width  -> Int, the width of the rectangle where the Ellipse will be drawn, is required
 height -> Int, the height of the rectangle where the Ellipse will be drawn, is required
 */
-(void)drawEllipse:(id)args;

/**
 Adds a clickable area into a rectangle which opens a link on click
 
 @param args (url, x, y, width, height)
 url    -> String, the link, is required
 x      -> Int, the top-down horizontal coordinate where 0 is the start point at top, is required
 y      -> Int, the left-right vertical coordinate where 0 is the start point at left, is required
 width  -> Int, the width of the rectangle for the clickable area, is required
 height -> Int, the height of the rectangle for the clickable area, is required
 */
-(void)addURL:(id)args;


/**
 Cancels the PDF drawing closing the context and deleting the PDF file

 @param args Void
 */
-(void)cancelPDF:(id)args;


/**
 Saves the PDF file closing the context and calling the js event "pdfReady"

 @param args (pdfName) PDF file name (without file extension)
 */
-(void)savePDF:(id)args;

@end


example/app.js
// open a single window
var win = Ti.UI.createWindow({
	backgroundColor:'white'
});
var label = Ti.UI.createLabel({
	text: "PDF Generator example"
});
win.add(label);
win.open();

var docViewer = Ti.UI.iOS.createDocumentViewer();
var pdfGenerator = require("br.com.grupow2abrasil.pdfgenerator");

pdfGenerator.addEventListener("pdfReady", function(e){
	Ti.API.info("url: " + e.url);
	docViewer.setUrl(e.url);
	docViewer.show();
});

// The PDF name doesn't need extension .pdf, only the name
pdfGenerator.setProperties({
	pdfName: "pdfOutputName",
    title: "The title of the PDF",
    author: "The author of the PDF",
    subject: "A subject if you want",
    keywords: "Keyworkds of the PDF",
    creator: "PDF Creator"
});

pdfGenerator.addNewPage();

pdfGenerator.drawText("The fox jumps over the lazy fox.", 100, 40, 200, 40);
pdfGenerator.drawLine(10, 10, 60, 60);

pdfGenerator.drawRect(5, 520, 100, 20);

// the image file must be in documents directory
var file = Ti.Filesystem.getFile(Titanium.Filesystem.getApplicationDataDirectory(), "yourFileName.jpg");

pdfGenerator.drawImage(file.nativePath, 500, 80, 16, 16);

pdfGenerator.setTextSize(12);
pdfGenerator.drawText("I'm a text with different font size", 10, 220, 300, 40);
pdfGenerator.setTextColor(158, 67, 89);
pdfGenerator.drawText("I'm a text with different color", 10, 120, 300, 40);

pdfGenerator.setLineWidth(5.0);
pdfGenerator.drawRect(50, 80, 25, 60);

pdfGenerator.savePDF();